src/damavand/cloud/aws/bucket.py

```
@@ -48,11 +51,63 @@ def add_object(self, object: bytes, path: str):
             )
         except ClientError as e:
             logger.error(f"Failed to add object to bucket `{self.name}`: {e}")
+            raise RuntimeError(e)
+
+    @runtime
+    def read(self, path: str) -> bytes:
+        try:
+            buffer = io.BytesIO()
+            self.__s3_client.download_fileobj(
+                Bucket=self.name, Key=path, Fileobj=buffer
+            )
+
+            return buffer.getvalue()
+        except ClientError as e:
+            logger.error(f"Failed to read object at `{path}`: {e}")
+            raise RuntimeError(e)
+
+    @runtime
+    def delete(self, path: str):
+        try:
+            self.__s3_client.delete_object(Bucket=self.name, Key=path)
+        except ClientError as e:
+            logger.error(f"Failed to delete object at `{path}`: {e}")
+            raise RuntimeError(e)
+
+    @runtime
+    def list(self) -> Iterable[str]:
+        """
+        Return an iterable of object keys in the bucket.
+
+        __ATTENTION__: This method is expensive for large buckets as it request multiple times to fetch all objects.
+        """
+        try:
+            paginator = self.__s3_client.get_paginator("list_objects_v2")
+            for page in paginator.paginate(Bucket=self.name):
+                for obj in page.get("Contents", []):
+                    yield obj["Key"]
+        except ClientError as e:
+            logger.error(f"Failed to list objects in storage `{self.name}`: {e}")
+            raise RuntimeError(e)
```

Raising the error, like this, will suppress the original stack trace. Can we make sure we maintain that? Also since this is a repeating pattern, maybe better to have a decorator for this error handling.

**PS:** not sure how useful it is to have `self.name` in errors, if it's not much, I'd rather to remove this error handling and let it raise asap.

---

@shahinism My idea was to unify the errors from ObjectStorage services regardless of the provider (aws or azure). This will for sure suppress the original stack trace but makes sure the end user is getting standard error type. My idea was to log the stack trace anyway (line 90) to make sure we won't miss any information.

---

@shahinism I took you advice and refactored. Could you please let me know what you think about this approch?

- the `logger.exception` in a try-catch will log the original exception with full information. However, the default raised exception unifies the exception for the end user.
- the match pattern will extract know edge-cases and translate them to meaningful and standard error classes to be handled by the end user.